sound/soc/sof/intel/hda.c

```diff
@@ -71,9 +71,19 @@ int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w)
 		return ret;
 	}
 
+	/* set flag and clear it after IPC */
+	set_mask_bits(&config->flags,
+		      SOF_DAI_CONFIG_FLAGS_MASK,
+		      SOF_DAI_CONFIG_FLAGS_HW_PARAMS);
+
 	/* send DAI_CONFIG IPC */
 	ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
 				  &reply, sizeof(reply));
+
+	set_mask_bits(&config->flags,
+		      SOF_DAI_CONFIG_FLAGS_MASK,
+		      SOF_DAI_CONFIG_FLAGS_NONE);
```

Comment:
can we drop this setting to `NONE` in both cases and instead add it where we're sending that value?

Reply:
Not following @lyakh. We need to set the flags, send the IPC and restore the flag. I don't get what you refer to with "where we're sending that value"? 

We would need to chase all cases where we send an DAI_CONFIG IPC, they are spread in multiple places so that would be more code overall.

Reply:
At the time of topology loading the config flags are always NONE. So, restoring the flags to NONE after sending the IPC makes sense.

Reply:
@ranj063 @plbossart I see it done in `sof_dai_config_setup()`, `hda_ctrl_dai_widget_setup()`, and `hda_ctrl_dai_widget_free()`, so, it's exactly one location apart from these two. Am I missing something?

Reply:
@lyakh its not the part about where it is sent that I'm concerned about. The DAI config data is saved. Why should we need to save a stale flag in the saved data. Isnt it better to clear it after using?

Reply:
@ranj063 it appears now that we have that configuration object that is cached for efficiency instead of being re-created for every use, right? Most of its data is static, some of it is dynamic but presents a "state," i.e. once its set it's valid until changed. And some of it is "single use" - it only makes sense for each specific use of that data. Basically that's this `.flags` field.
Clearly data from the first category is set once and never changed. From the second one is updated every time it changes and is valid at any time. As for the third one - you're suggesting to keep it in some "default state" and only change when needed, whereas I'm proposing to explicitly set it every time before use. It isn't hugely important of course, both work, yours is potentially safer whereas mine forces the user to explicitly set valid values. We could in fact even make the default some invalid value, like `~0`. So after use we'd always set all such volatile fields to their default possibly with `memset(0xff)` and in the firmware we could check for that value and complain. That way we'd have the best of both worlds - no stale data and enforced validity. But I'm probably overengineering :-) Don't think it's important enough.

Reply:
I agree with the analysis that the addition of the flags changes the dai_config structure from a static values to partially dynamic ones. that's a fair point.

I'll update the code to follow @lyakh 's representation, my initial proposal was based on ignorance of where the IPC is send. We should have searchable commands for all IPCs so that we can find out what happens where, using config->hdr.cmd is not wrong but makes it hard for the average developer to find the source of an IPC.